README.md
Scorpio SphinxSearchBundle
==========================

SphinxSearchBundle adds configuration and service support to Scorpio SphinxSearch,
making it easier to use in a Laravel project.

Requirements
------------

 * Laravel 5.2+
 * Scorpio SphinxSearch
 * for composer installs, PHP Sphinx extension

Installation
------------

 1. The preferred method is to install via composer:

    composer require scorpio/sphinx-search-laravel "0.1.*"

 2. Enable the service provider in your config/app.php:

    // config/app.php
    \Scorpio\SphinxSearchLaravel\SphinxSearchServiceProvider::class

 3. ./artisan vendor:publish

 4. Set the configuration parameters in config/sphinx.php

 5. And map some indexes (see later)

Basic Usage
-----------

This bundle exposes the following configuration:

    host: localhost
    port: 9312
    max_query_time: 5000 # max query execution time

Optionally a specific SphinxClient class can be specified to handle the connections.
This can be used if the PHP extension is not available and the SphinxQL library
cannot be used.

    client_class: SomeClass\That\Implements\SphinxClientAPI::class

The following services are automatically registered:

 * scorpio_sphinx_search.server.settings (server settings)
 * scorpio_sphinx_search.search_manager  (main search manager instance)

Indexes can be defined in the config file and will be published as services:

```php
'indexes' => [
    'my_custom_sphinx_index' => [
        'class'     => \Scorpio\SphinxSearch\SearchIndex::class,
        'arguments' => [
            'index'        => 'my_custom_sphinx_index',
            'fields'       => [ 'available', 'fields', 'as_an_array' ],
            'attributes'   => [ 'attribute1', 'attribute2' ],
        ],
    ],
],
```

Note: the index name and fields are required and must match what is exposed in the
Sphinx configuration.

Additionally the result set and result record class can also be specified:

```php
'indexes' => [
    'my_custom_sphinx_index' => [
        'class'     => \Scorpio\SphinxSearch\SearchIndex::class,
        'arguments' => [
            'index'        => 'my_custom_sphinx_index',
            'fields'       => [ 'available', 'fields', 'as_an_array' ],
            'attributes'   => [ 'attribute1', 'attribute2' ],
            'result_set'   => MyResultSet::class,
            'result_class' => MyCustomResult::class,
        ],
    ],
],
```

Finally, for the really lazy!, the index can have a 'query' element added:

```php
'indexes' => [
    'my_custom_sphinx_index' => [
        'class'     => \Scorpio\SphinxSearch\SearchIndex::class,
        'query'     => true,
        'arguments' => [
            'index'        => 'my_custom_sphinx_index',
            'fields'       => [ 'available', 'fields', 'as_an_array' ],
            'attributes'   => [ 'attribute1', 'attribute2' ],
        ],
    ],
],
```

A custom query service will be automatically registered in the container. The prefix
can be customised in your config file with the default being "query.", so the
previous example would create the service: "query.my_custom_sphinx_index".

Note: the attribute "query" must be set to true, otherwise the index will be ignored.
This allows the services to be tagged and locatable for debugging but not auto-create
a query service when not needed.

In your controller you can then access the query instance:

```php
class MyController extends Controller
{

    function indexAction(Request $request)
    {
        // bind a search term somehow, apply filters etc. maybe check for keywords...
        $query = app('query.my_custom_sphinx_index')
            ->setQuery($request->query->get('keywords'));

        $results = app('scorpio_sphinx_search.search_manager')->query($query);

        // do something with the results.
    }
}
```

For Laravel, you are encouraged to create custom SearchIndex classes for each index,
and to then name then using the class name. Then they can be injected as dependencies.

License
-------

This library is licensed under the BSD license. See the complete license in the included
LICENSE file.

Issues or feature requests
---------------------------

Issues and feature requests should be made on the [Github repository page](https://github.com/scorpioframework/sphinx-search-laravel/issues).


CHANGELOG.md
Change Log
==========

2017-09-02
----------

 * Updated dependencies

2017-02-03
----------

 * Updated dependencies for Laravel 5.4
 
2015-11-26
----------

Initial commit.


src/SphinxSearchServiceProvider.php
<?php

/*
 * This file is part of the Scorpio SphinxSearch Laravel Bundle.
 *
 * (c) Dave Redfern <dave@scorpioframework.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Scorpio\SphinxSearchLaravel;

use Illuminate\Support\ServiceProvider;
use Scorpio\SphinxSearch\SearchManager;
use Scorpio\SphinxSearch\SearchQuery;
use Scorpio\SphinxSearch\ServerSettings;

/**
 * Class TenancyServiceProvider
 *
 * @package    Somnambulist\Tenancy\Providers
 * @subpackage Somnambulist\Tenancy\Providers\TenancyServiceProvider
 * @author     Dave Redfern
 */
class SphinxSearchServiceProvider extends ServiceProvider
{

    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot()
    {
        $this->publishes([ $this->getConfigPath() => config_path('sphinx.php'), ], 'config');
    }

    /**
     * Register the service provider.
     *
     * @return void
     */
    public function register()
    {
        $this->mergeConfig();

        $this->registerServerSettings();
        $this->registerSearchQueryProvider();
        $this->registerIndexDefinitions();
    }



    /**
     * Merge config
     */
    protected function mergeConfig()
    {
        $this->mergeConfigFrom($this->getConfigPath(), 'sphinx');
    }

    /**
     * Register the Sphinx server settings instance
     *
     * @return void
     */
    protected function registerServerSettings()
    {
        $this->app->singleton('scorpio_sphinx_search.server.settings', function ($app) {
            $config = $app->make('config');

            if ( !extension_loaded('sphinx') && !$config->get('sphinx.client_class') ) {
                throw new \RuntimeException(
                    sprintf('SphinxClientAPI is not available and a client_class was not defined')
                );
            }

            return new ServerSettings(
                $config->get('sphinx.host'),
                $config->get('sphinx.port'),
                $config->get('sphinx.max_query_time'),
                $config->get('sphinx.client_class')
            );
        });
    }

    /**
     * Registers the search manager that holds the indexes
     *
     * @return void
     */
    protected function registerSearchQueryProvider()
    {
        $this->app->singleton('scorpio_sphinx_search.search_query_provider', function ($app) {
            return new SearchManager($app['scorpio_sphinx_search.server.settings']);
        });
    }

    /**
     * Register the indexes
     *
     * @return void
     */
    protected function registerIndexDefinitions()
    {
        /** @var SearchManager $provider */
        $provider = $this->app->make('scorpio_sphinx_search.search_query_provider');
        $config   = $this->app->make('config');
        $prefix   = $config->get('sphinx.query_search_prefix');

        foreach ($config->get('sphinx.indexes', []) as $service => $details) {
            $this->app->singleton($service, function ($app) use ($details) {
                $class  = $details['class'];
                $args   = $details['arguments'];
                $set    = isset($args['result_set']) ? $args['result_set'] : null;
                $result = isset($args['result_class']) ? $args['result_class'] : null;

                return new $class($args['index'], $args['fields'], $args['attributes'], $set, $result);
            });

            if (isset($details['query']) && $details['query']) {
                $this->app->instance($prefix . $service, new SearchQuery($this->app[$service]));
                $provider->addQuery($this->app[$prefix . $service]);
            }
        }
    }

    /**
     * @return string
     */
    protected function getConfigPath()
    {
        return __DIR__ . '/../config/sphinx.php';
    }

    /**
     * @return array
     */
    public function provides()
    {
        return [
            'scorpio_sphinx_search.server.settings',
            'scorpio_sphinx_search.search_query_provider',
        ];
    }
}

config/sphinx.php
<?php

return [

    /*
     * The host for sphinx server
     */
    'host' => 'localhost',

    /*
     * The port
     */
    'port' => 9312,

    /*
     * Maximum execution time in MILLIseconds that a query should run.
     * 0 is infinite, minimum should be around 5000.
     */
    'max_query_time' => 5000,

    /*
     * Optional: a class that implements the SphinxClientAPI if not using the sphinx extension.
     */
    'client_class' => null,

    /*
     * The prefix for all auto-created query
     */
    'query_service_prefix' => 'query.',

    /*
     * Add your configured indexes here, they will be loaded and made available as services
     * within the container. The format is similar to the Symfony container yaml.
     *
     * Each index has:
     *  * a unique name that will be used as the service name,
     *  * a class that implements the index,
     *  * the arguments that define the index:
     *    * index - the actual index name that is defined in the sphinx config
     *    * fields - the fields that are available in the index
     *    * attributes - the attributes that can be filtered on
     *    * result_set - a class name for a custom result set (optional)
     *    * result_class - a class name for a custom result record (optional)
     *  * query - if true, a query instance will be created avoiding the need to set that bit up.
     *
     * Query instances are prefixed with query. by default (change above). So "service_name_for_index"
     * would create: "query.service_name_for_index" as the service in the container.
     */
    'indexes' => [

        'service_name_for_index' => [
            'class'     => \Scorpio\SphinxSearch\SearchIndex::class,
            //'query'     => true,
            'arguments' => [
                'index'        => 'index_name',
                'fields'       => [],
                'attributes'   => [],
                //'result_set'   => null,
                //'result_class' => null,
            ],
        ]
    ],

];